# Lab 4: Neural Networks #


In [ ]:
import pandas as pd
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVC

from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.feature_extraction import DictVectorizer

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV, ParameterGrid

import numpy as np

import warnings
warnings.filterwarnings("ignore")

In [ ]:
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
!wget http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv

df_train = pd.read_csv('./lab_4_training.csv')
df_test = pd.read_csv('./lab_4_test.csv')
df_train.head()

--2019-10-01 06:25:08--  http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_training.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 105581 (103K) [text/csv]
Saving to: ‘lab_4_training.csv.11’

lab_4_training.csv. 100%[===================>] 103.11K   487KB/s    in 0.2s    

2019-10-01 06:25:08 (487 KB/s) - ‘lab_4_training.csv.11’ saved [105581/105581]

--2019-10-01 06:25:09--  http://people.ischool.berkeley.edu/~zp/course_datasets/lab_4_test.csv
Resolving people.ischool.berkeley.edu (people.ischool.berkeley.edu)... 128.32.78.16
Connecting to people.ischool.berkeley.edu (people.ischool.berkeley.edu)|128.32.78.16|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 26523 (26K) [text/csv]
Saving to: ‘lab_4_test.csv.11’

lab_4_test.csv.11   100%[===================>

Unnamed: 0 gender  age    year  ... exercisehours  musiccds  playgames  watchtv
0        1303   male   20  second  ...           5.0      50.0        1.0     15.0
1          36   male   20   third  ...           4.0      10.0        0.0      1.0
2         489   male   22  fourth  ...           2.0     150.0        1.0     10.0
3        1415   male   19  second  ...           5.0     100.0        0.0      7.0
4         616   male   22  fourth  ...           7.0      10.0        0.0      5.0

[5 rows x 15 columns]

***
### Question 1###
Calculate a baseline accuracy measure using the majority class, assuming a target variable of 'gender'. The majority class is the most common value of the target variable in a particular dataset. Accuracy is calculated as (true positives + true negatives) / (all negatives and positives)

**Question 1.a**  
Find the majority class in the training set. If you always predicted this class in the training set, what would your accuracy be?

In [ ]:
# YOUR CODE HERE
df_train.groupby('gender').count()
majority = "female"
true_positive = (df_train['gender'] == 'female').sum()
true_negative = 0
total = df_train['gender'].count()

accuracy = (true_positive + true_negative) / total
accuracy

0.5377358490566038

###ANSWER: 0.5377358490566038

**Question 1.b**   
If you always predicted this same class (majority from the training set) in the test set, what would your accuracy be?

In [ ]:
# YOUR CODE HERE
df_test.groupby('gender').count()
majority_test = "female"
true_positive_test = (df_test['gender'] == 'female').sum()
true_negative_test = 0
total_test = df_test['gender'].count()

accuracy_test = (true_positive_test + true_negative_test) / total_test
accuracy_test

0.5226130653266332

###ANSWER: 0.5226130653266332

***
### Question 2 ###
Get started with Neural Networks.

   
Choose a NN implementation (eg: scikit-learn) and specify which you choose. Be sure the implementation allows you to modify the number of hidden layers and hidden nodes per layer.  

NOTE: When possible, specify the logsig (sigmoid/logistc) function as the transfer function (another word for activation function) and use Levenberg-Marquardt backpropagation (lbfgs). It is possible to specify logistic in Sklearn MLPclassifier (Neural net).  

**Question 2.a**   
Train a neural network with a single 10 node hidden layer. Only use the Height feature of the dataset to predict the Gender. You will have to change Gender to a 0 and 1 class. After training, use your trained model to predict the class using the height feature from the training set. What was the accuracy of this prediction?

NN implementation: scikit-learn

In [ ]:
df_train = df_train.replace({'first"': "first"})
df_test = df_test.replace({'first"': "first"})

In [ ]:
df_train.head()
r_df_train = df_train.replace('male', 0).replace('female', 1)
r_df_test = df_test.replace('male', 0).replace('female', 1)


In [ ]:
# YOUR CODE HERE

X_train = r_df_train[['height']]
Y_train = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=100, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)

clf.fit(X_train,Y_train)
print('Accuracy on training---')
y_pred_train=clf.predict(X_train)
print(accuracy_score(Y_train,y_pred_train))

Accuracy on training---
0.539622641509434


###ANSWER: 0.539622641509434

**Question 2.b**  
Take the trained model from question 2.a and use it to predict the test set. This can be accomplished by taking the trained model and giving it the Height feature values from the test set. What is the accuracy of this model on the test set?

In [ ]:
# YOUR CODE HERE

X_test = r_df_test[['height']]
Y_test = r_df_test['gender']

print('Accuracy on test---')
y_pred_test=clf.predict(X_test)
print(accuracy_score(Y_test,y_pred_test))

Accuracy on test---
0.5226130653266332


###ANSWER: 0.5226130653266332

**Question 2.c**   
Neural Networks tend to prefer smaller, normalized feature values. Try taking the log of the height feature in both training and testing sets or use a Standard Scalar operation in SKlearn to centre and normalize the data between 0-1 for continuous values. Repeat question 2.a and 2.b with the log version or the normalized and centered version of this feature

In [ ]:
# YOUR CODE HERE
log_X_train = np.log(r_df_train[['height']])
log_Y_train = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=100, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)

clf.fit(log_X_train,log_Y_train)
print('Accuracy on training---')
log_y_pred_train=clf.predict(log_X_train)
print(accuracy_score(log_Y_train,log_y_pred_train))

log_X_test = np.log(r_df_test[['height']])
log_Y_test = r_df_test['gender']

print('Accuracy on test---')
log_y_pred_test=clf.predict(log_X_test)
print(accuracy_score(log_Y_test,log_y_pred_test))

Accuracy on training---
0.6842767295597484
Accuracy on test---
0.6683417085427136


###ANSWER: Accuracy on training: 0.6842767295597484; Accuracy on test: 0.6683417085427136

***

### Question 3###
The rest of features in this dataset barring a few are categorical. No ML method accepts categorical features, so transform year, eyecolor, exercise into a set of binary features, one feature per unique original feature value, and mark the binary feature as ‘1’ if the feature value matches the original value and ‘0’ otherwise. Using only these binary variable transformed features, train and predict the class of the test set. What was your accuracy using Neural Network with a single 10 node hidden layer? During training, use a maximum number of iterations of 50.

In [ ]:
# YOUR CODE HERE
from sklearn import preprocessing
df_train

def encoded(r_df_train):
  cat_df = r_df_train.select_dtypes(include=[object])
  le = preprocessing.LabelEncoder()
  cat_df_2 = cat_df.apply(le.fit_transform)
  enc = preprocessing.OneHotEncoder()
  enc.fit(cat_df_2)
  onehotlabels = enc.transform(cat_df_2).toarray()
  encoded_cat_df = pd.DataFrame(onehotlabels)
  return encoded_cat_df


#training
cat_X_train = encoded(r_df_train)
cat_Y_train = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)

clf.fit(cat_X_train,cat_Y_train)
print('Accuracy on training---')
cat_y_pred_train=clf.predict(cat_X_train)
print(accuracy_score(cat_Y_train,cat_y_pred_train))

cat_X_test = encoded(r_df_test)
cat_Y_test = r_df_test['gender']

print('Accuracy on test---')
cat_y_pred_test=clf.predict(cat_X_test)
print(accuracy_score(cat_Y_test,cat_y_pred_test))

Accuracy on training---
0.5735849056603773
Accuracy on test---
0.6080402010050251


###ANSWER: Accuracy on training: 0.5735849056603773; Accuracy on test: 0.6080402010050251

***
### Question 4###
Using a NN, report the accuracy on  the test set of a model that trained only on the height and the eye color features of instances in the training set.

**Question 4.a**  
What is the accuracy on the test set using the original height values (no pre-processing) and eye color as a one-hot?

In [ ]:
# YOUR CODE HERE
df_4 = r_df_train[['height', 'eyecolor']]
df_4_test = r_df_test[['height', 'eyecolor']]
encoded_4 = encoded(df_4)
encoded_4['height'] = df_4['height']
encoded_4_test = encoded(df_4_test)
encoded_4_test['height'] = df_4_test['height']

#training
e_X_train = encoded_4
e_Y_train = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)
clf.fit(e_X_train,e_Y_train)
print('Accuracy on training---')
e_y_pred_train=clf.predict(e_X_train)
print(accuracy_score(e_Y_train,e_y_pred_train))

e_X_test = encoded_4_test
e_Y_test = r_df_test['gender']

print('Accuracy on test---')
e_y_pred_test=clf.predict(e_X_test)
print(accuracy_score(e_Y_test,e_y_pred_test))

Accuracy on training---
0.5849056603773585
Accuracy on test---
0.5954773869346733


###ANSWER: Accuracy on training: 0.5849056603773585; Accuracy on test: 0.5954773869346733

**Question 4.b**  
What is the accuracy on the test set using the log of height values (applied to both training and testing sets) and eye color as a one-hot?

In [ ]:
# YOUR CODE HERE
df_4b = r_df_train[['height', 'eyecolor']]
df_4b_test = r_df_test[['height', 'eyecolor']]
encoded_4b = encoded(df_4b)
encoded_4b['height'] = np.log(df_4b['height'])
encoded_4b_test = encoded(df_4b_test)
encoded_4b_test['height'] = np.log(df_4b_test['height'])

#training
be_X_train = encoded_4b
be_Y_train = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)
clf.fit(be_X_train,be_Y_train)
print('Accuracy on training---')
be_y_pred_train=clf.predict(be_X_train)
print(accuracy_score(be_Y_train,be_y_pred_train))

be_X_test = encoded_4b_test
be_Y_test = r_df_test['gender']

print('Accuracy on test---')
be_y_pred_test=clf.predict(be_X_test)
print(accuracy_score(be_Y_test,be_y_pred_test))

Accuracy on training---
0.6742138364779874
Accuracy on test---
0.6708542713567839


###ANSWER: Accuracy on training: 0.6742138364779874; Accuracy on test: 0.6708542713567839

**Question 4.c**  
What is the accuracy on the test set using the Z-score of height values and eye color as a one-hot? 

Z-score is a normalization function. It is the value of a feature minus the average value for that feature (in the training set), divided by the standard deviation of that feature (in the training set). Remember that, whenever applying a function to a feature in the training set, it also has to be applied to that same feature in the test set.

In [ ]:
# YOUR CODE HERE
df_4c = r_df_train[['height', 'eyecolor']]
df_4c_test = r_df_test[['height', 'eyecolor']]
encoded_4c = encoded(df_4c)
encoded_4c_test = encoded(df_4c_test)

encoded_4c['height'] = (df_4c['height'] - df_4c['height'].mean())/df_4c['height'].std()
encoded_4c_test['height'] = (df_4c_test['height'] - df_4c_test['height'].mean())/df_4c_test['height'].std()

#training
ce_X_train = encoded_4c
ce_Y_train = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)
clf.fit(ce_X_train,ce_Y_train)
print('Accuracy on training---')
ce_y_pred_train=clf.predict(ce_X_train)
print(accuracy_score(ce_Y_train,ce_y_pred_train))

ce_X_test = encoded_4c_test
ce_Y_test = r_df_test['gender']

print('Accuracy on test---')
ce_y_pred_test=clf.predict(ce_X_test)
print(accuracy_score(ce_Y_test,ce_y_pred_test))

Accuracy on training---
0.8534591194968554
Accuracy on test---
0.8693467336683417


###ANSWER: Accuracy on training: 0.8534591194968554; Accuracy on test: 0.8693467336683417

***
### Question 5 ###
Repeat question 4 for exercise hours + eye color

**Question 5.a**  
What is the accuracy on the test set using the original exercise hours (no pre-processing) and eye color as a one-hot?

In [ ]:
# YOUR CODE HERE

df_5 = r_df_train[['exercisehours', 'eyecolor']]
df_5_test = r_df_test[['exercisehours', 'eyecolor']]
encoded_5 = encoded(df_5)
encoded_5['exercisehours'] = df_5['exercisehours']
encoded_5_test = encoded(df_5_test)
encoded_5_test['exercisehours'] = df_5_test['exercisehours']

#training
e_X_train5 = encoded_5
e_Y_train5 = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)
clf.fit(e_X_train5,e_Y_train5)
print('Accuracy on training---')
e_y_pred_train5=clf.predict(e_X_train5)
print(accuracy_score(e_Y_train5,e_y_pred_train5))

e_X_test5 = encoded_5_test
e_Y_test5 = r_df_test['gender']

print('Accuracy on test---')
e_y_pred_test5=clf.predict(e_X_test5)
print(accuracy_score(e_Y_test5,e_y_pred_test5))

Accuracy on training---
0.5817610062893082
Accuracy on test---
0.5678391959798995


###ANSWER: Accuracy on training: 0.5817610062893082; Accuracy on test: 0.5678391959798995

**Question 5.b**  
What is the accuracy on the test set using the log of exercise values (applied to both training and testing sets) and eye color as a one-hot?

In [ ]:

data_train = r_df_train[['exercisehours', 'eyecolor', 'gender']]
data_test = r_df_test[['exercisehours', 'eyecolor', 'gender']]

data_train['exercisehours'] = data_train['exercisehours'].replace(0.0, 0.01)
data_test['exercisehours'] = data_test['exercisehours'].replace(0.0,0.01)

df_5b = data_train[['exercisehours', 'eyecolor']]
df_5b_test = data_test[['exercisehours', 'eyecolor']]

encoded_5b = encoded(df_5b)
encoded_5b['exercisehours'] = np.log(df_5b['exercisehours'])

encoded_5b_test = encoded(df_5b_test)
encoded_5b_test['exercisehours'] = np.log(df_5b_test['exercisehours'])

#training
X_train = encoded_5b
Y_train = data_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)
clf.fit(X_train,Y_train)
print('Accuracy on training---')
y_pred_train=clf.predict(X_train)
print(accuracy_score(Y_train,y_pred_train))

X_test = encoded_5b_test
Y_test = data_test['gender']

print('Accuracy on test---')
y_pred_test=clf.predict(X_test)
print(accuracy_score(Y_test,y_pred_test))


Accuracy on training---
0.5830188679245283
Accuracy on test---
0.5728643216080402


###ANSWER: Accuracy on training: 0.5830188679245283; Accuracy on test: 0.5728643216080402

In [ ]:
# YOUR CODE HERE
df_5c = r_df_train[['exercisehours', 'eyecolor']]
df_5c_test = r_df_test[['exercisehours', 'eyecolor']]
encoded_5c = encoded(df_5c)
encoded_5c_test = encoded(df_5c_test)

encoded_5c['exercisehours'] = (df_5c['exercisehours'] - df_5c['exercisehours'].mean())/df_5c['exercisehours'].std()
encoded_5c_test['exercisehours'] = (df_5c_test['exercisehours'] - df_5c_test['exercisehours'].mean())/df_5c_test['exercisehours'].std()

#training
ce_X_train5 = encoded_5c
ce_Y_train5 = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)
clf.fit(ce_X_train5,ce_Y_train5)
print('Accuracy on training---')
ce_y_pred_train5=clf.predict(ce_X_train5)
print(accuracy_score(ce_Y_train5,ce_y_pred_train5))

ce_X_test5 = encoded_5c_test
ce_Y_test5 = r_df_test['gender']

print('Accuracy on test---')
ce_y_pred_test5=clf.predict(ce_X_test5)
print(accuracy_score(ce_Y_test5,ce_y_pred_test5))

Accuracy on training---
0.5867924528301887
Accuracy on test---
0.5728643216080402


###ANSWER: Accuracy on training: 0.5867924528301887; Accuracy on test: 0.5728643216080402

***
### Question 6###
Combine the features from question 3, 4, and 5 (year, eyecolor, exercise, height, exercise hours). For numeric features use the best normalization method from questions 4 and 5.

**Question 6.a**  
What was the NN accuracy on the test set using the single 10 node hidden layer?

In [ ]:
# YOUR CODE HERE
df_6_train = encoded(r_df_train)
df_6_train['height'] = (r_df_train['height'] - r_df_train['height'].mean())/r_df_train['height'].std()
df_6_train['exercisehours'] = (r_df_train['exercisehours'] - r_df_train['exercisehours'].mean())/r_df_train['exercisehours'].std()

df_6_test = encoded(r_df_test)
df_6_test['height'] = (r_df_test['height'] - r_df_test['height'].mean())/r_df_test['height'].std()
df_6_test['exercisehours'] = (r_df_test['exercisehours'] - r_df_test['exercisehours'].mean())/r_df_test['exercisehours'].std()

cat_X_train = df_6_train
cat_Y_train = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,), activation = 'logistic', max_iter=50, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)

clf.fit(cat_X_train,cat_Y_train)
print('Accuracy on training---')
cat_y_pred_train=clf.predict(cat_X_train)
print(accuracy_score(cat_Y_train,cat_y_pred_train))

cat_X_test = df_6_test
cat_Y_test = r_df_test['gender']

print('Accuracy on test---')
cat_y_pred_test=clf.predict(cat_X_test)
print(accuracy_score(cat_Y_test,cat_y_pred_test))


Accuracy on training---
0.8553459119496856
Accuracy on test---
0.864321608040201


###ANSWER: Accuracy on training: 0.8553459119496856; Accuracy on test: 0.864321608040201

***
### Question 7- Bonus (10%)###
Can you improve your test set prediction accuracy by 5% or more?  

See how close to that milestone of improvement you can get by modifying the tuning parameters of  Neural Networks (the number of hidden layers, number of hidden nodes in each layer, the learning rate aka mu). A great guide to tuning parameters is explained in this guide: http://www.csie.ntu.edu.tw/~cjlin/papers/guide/guide.pdf. 

While the guide is specific to SVM and in particular the C and gamma parameters of the RBF kernel, the method applies to generally to any ML technique with tuning parameters.

Please also write a paragraph in a markdown cell below with an explanation of your approach and evaluation metrics.


In [ ]:
# YOUR CODE HERE
df_7_train = encoded(r_df_train)
df_7_train['height'] = (r_df_train['height'] - r_df_train['height'].mean())/r_df_train['height'].std()
df_7_train['exercisehours'] = (r_df_train['exercisehours'] - r_df_train['exercisehours'].mean())/r_df_train['exercisehours'].std()

df_7_test = encoded(r_df_test)
df_7_test['height'] = (r_df_test['height'] - r_df_test['height'].mean())/r_df_test['height'].std()
df_7_test['exercisehours'] = (r_df_test['exercisehours'] - r_df_test['exercisehours'].mean())/r_df_test['exercisehours'].std()

cat_X_train7 = df_7_train
cat_Y_train7 = r_df_train['gender']
clf = MLPClassifier(hidden_layer_sizes=(10,10,), activation = 'logistic', max_iter=100, alpha=0.0001,
                     solver='lbfgs', verbose=10,  random_state=21,tol=0.000000001)

clf.fit(cat_X_train7,cat_Y_train7)
print('Accuracy on training---')
cat_y_pred_train7=clf.predict(cat_X_train7)
print(accuracy_score(cat_Y_train7,cat_y_pred_train7))

cat_X_test7 = df_7_test
cat_Y_test7 = r_df_test['gender']

print('Accuracy on test---')
cat_y_pred_test7=clf.predict(cat_X_test7)
print(accuracy_score(cat_Y_test7,cat_y_pred_test7))

Accuracy on training---
0.8647798742138365
Accuracy on test---
0.871859296482412


###ANSWER: My changing the hidden layer sizes from just 10 to 10,10, the accuracy increased. In addition, changing the number of iterations also changed the accuracy of the test.